plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderPortComparator.java

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Kiel University and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.elk.alg.layered.intermediate.preserveorder;
+
+import java.util.Comparator;
+import java.util.Map;
+
+import org.eclipse.elk.alg.layered.graph.LEdge;
+import org.eclipse.elk.alg.layered.graph.LNode;
+import org.eclipse.elk.alg.layered.graph.LNode.NodeType;
+import org.eclipse.elk.alg.layered.graph.LPort;
+import org.eclipse.elk.alg.layered.graph.Layer;
+import org.eclipse.elk.alg.layered.options.InternalProperties;
+
+/**
+ * Orders {@link LPort}s in the same layer by the of their edges {@link InternalProperties#MODEL_ORDER}
+ * or the order of the nodes they connect to in the previous layer.
+ * Outgoing ports are ordered before incoming ports.
+ * Incoming ports choose a position that does not create conflicts with the previous layer.
+ */
+public class ModelOrderPortComparator implements Comparator<LPort> {
+
+    /**
+     * The model order associated to a target node of this node.
+     */
+    private final Map<LNode, Integer> targetNodeModelOrder;
+
+    /**
+     * The previous layer.
+     */
+    private final Layer previousLayer;
+
+    /**
+     * Creates a comparator to compare {@link LPort}s in the same layer.
+     * 
+     * @param previousLayer The previous layer
+     * @param targetNodeModelOrder The minimal model order connecting to a target node.
+     */
+    public ModelOrderPortComparator(final Layer previousLayer, final Map<LNode, Integer> targetNodeModelOrder) {
+        this.previousLayer = previousLayer;
+        this.targetNodeModelOrder = targetNodeModelOrder;
+    }
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see java.util.Comparator#compare(java.lang.Object, java.lang.Object)
+     */
+    @Override
+    public int compare(final LPort p1, final LPort p2) {
+        // Sort incoming edges by sorting their ports by the order of the nodes they connect to.
+        if (p1.getOutgoingEdges().isEmpty() && p2.getOutgoingEdges().isEmpty()) {
+            LNode p1Node = p1.getIncomingEdges().get(0).getSource().getNode();
```

Comment:
Is it guaranteed that an incoming edge exists? If so, could you add a comment stating this fact?

Reply:
I guess not, maybe it would be better to check for non empty than for empty.